### PR TITLE
Add a replacement mechanism to work around nil handling in ucfg

### PR DESCRIPTION
**Pull request by pebrc**
Stopping short of replacing our use of ucfg alltogether this tries to provide a short term fix to the rendering problem that currently prevents users specifying Elasticsearch coordinator nodes using the new `node.roles: []` syntax. 

This introduces a simple (hacky) replacement mechanism into the configuration rendering process to handle the cases where we know that null is not equivalent to [] (empty seq) in the rendered Yaml configuration.

If we think this is viable as short term fix for https://github.com/elastic/cloud-on-k8s/issues/3718 I would add a few more tests and see to get this merged for 1.4.0

### CODE REVIEW

**Review by sebgl — COMMENTED** [review]
I think I'm good with the approach 👍 
Let's make sure we add a comment to explain why we do this.

(I fiddle a bit with patching the raw user input instead of the config rendering but it does not really make things easier...)

File: pkg/controller/common/settings/replace.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+				replaceTarget[p] = r.Replacement
+				return
+			}
+		}
+
+		v, exists := replaceTarget[p].(untypedDict)
+		if !exists {
+			return
+		}
+		replaceTarget = v
```
Comment (sebgl) [inline pkg/controller/common/settings/replace.go:33]: The 4 lines above give me a headache for some reason 😄 
we mutate `replaceTarget` by the inner path value here, right? A comment would help, but also variable names a bit more explicit than `v` and `p`.

Comment (pebrc) [reply pkg/controller/common/settings/replace.go]: You are right, it is too terse. I revised the naming.

File: pkg/apis/elasticsearch/v1/elasticsearch_config.go [hunk 2]
```diff
@@ -37,13 +39,14 @@ type ClusterSettings struct {
...
-	ML         *bool    `config:"ml"`
-	Transform  *bool    `config:"transform"` // available as of 7.7.0
-	Roles      []string `config:"roles"`     // available as of 7.9.0, takes priority over the other fields if non-nil
-	VotingOnly *bool    `config:"voting_only"`
+	Master              *bool    `config:"master"`
+	Data                *bool    `config:"data"`
+	Ingest              *bool    `config:"ingest"`
+	ML                  *bool    `config:"ml"`
+	Transform           *bool    `config:"transform"`             // available as of 7.7.0
+	RemoteClusterClient *bool    `config:"remote_cluster_client"` // available as of 7.7.0
```
Comment (pebrc) [inline pkg/apis/elasticsearch/v1/elasticsearch_config.go:47]: We have an inconsistency with our sset pod labelling. We label the `transform` role but we don't label the `voting_only` or the `remote_cluster_role` (which we ignored until this PR). I am not sure what to do about it. Are there node roles that are more important than others? If we consider them all equals we should add labels for all of them on the pods.

File: pkg/controller/common/settings/replace_test.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+					"something": "else",
+				},
+			},
+			expected: []byte(`node:
+  something: else
+`),
+		},
+	}
+
+	for _, tt := range tests {
```
Comment (david-kow) [inline pkg/controller/common/settings/replace_test.go:64]: Maybe invoke inside `t.Run`?

File: pkg/controller/common/settings/replace_test.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	tests := []struct {
+		name     string
+		input    map[string]interface{}
+		expected []byte
+	}{
+		{
+			name: "nil roles replacement",
+			input: map[string]interface{}{
+				"node": map[string]interface{}{
+					"roles": make([]string, 0),
```
Comment (david-kow) [inline pkg/controller/common/settings/replace_test.go:32]: Should this be `roles: nil`? Or maybe add a separate test to validate if the "[] to nil" behavior is still there in ucfg?

File: test/e2e/test/elasticsearch/checks_es.go [hunk 5]
```diff
@@ -186,24 +186,33 @@ func (e *esClusterChecks) CheckESNodesTopology(es esv1.Elasticsearch) test.Step
 }
 
 func compareRoles(expected *esv1.Node, actualRoles []string) bool {
```
[the remark this answers is not in the document]
Comment (pebrc) [reply test/e2e/test/elasticsearch/checks_es.go:188]: Well spotted!

File: pkg/apis/elasticsearch/v1/elasticsearch_config.go (hunk 2 above)
Comment (sebgl) [reply pkg/apis/elasticsearch/v1/elasticsearch_config.go:47]: I think I'm +1 for labelling them all.

Comment (pebrc) [reply pkg/apis/elasticsearch/v1/elasticsearch_config.go:47]: OK will add another commit here.

Comment (pebrc) [reply pkg/apis/elasticsearch/v1/elasticsearch_config.go:47]: I added a commit only to realise that there is more amiss #4054

File: pkg/controller/common/settings/replace_test.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+			expected: []byte(`node:
+  roles:
+  - master
+`),
+		},
+		{
+			name: "does not touch other paths",
+			input: map[string]interface{}{
+				"node": map[string]interface{}{
+					"something": "else",
```
Comment (sebgl) [inline pkg/controller/common/settings/replace_test.go:68]: would be more interesting if it was `"something": nil`

File: pkg/controller/elasticsearch/label/label.go [hunk 7]
```diff
@@ -42,6 +42,10 @@ const (
 	NodeTypesMLLabelName common.TrueFalseLabel = "elasticsearch.k8s.elastic.co/node-ml"
 	// NodeTypesTransformLabelName is a label set to true on nodes with the transform role
 	NodeTypesTransformLabelName common.TrueFalseLabel = "elasticsearch.k8s.elastic.co/node-transform"
+	//NodeTypesRemoteClusterClientLabelName is a label set to true on nodes with the remote_cluster_client role
+	NodeTypesRemoteClusterClientLabelName common.TrueFalseLabel = "elasticsearch.k8s.elastic.co/node-remote_cluster_client"
+	//NodeTypesVotingOnlyLabelName is a label set to true on nodes with voting_only master-eligible node
```
Comment (sebgl) [inline pkg/controller/elasticsearch/label/label.go:47]: nit: `//<whitespace>`

File: pkg/controller/elasticsearch/label/label.go [hunk 8]
```diff
@@ -116,9 +120,14 @@ func NewPodLabels(
 	NodeTypesDataLabelName.Set(nodeRoles.HasDataRole(), labels)
 	NodeTypesIngestLabelName.Set(nodeRoles.HasIngestRole(), labels)
 	NodeTypesMLLabelName.Set(nodeRoles.HasMLRole(), labels)
-	// transform nodes were only added in 7.7.0 so we should not annotate previous versions with them
+	// transform and remote_cluster_client roles were only added in 7.7.0 so we should not annotate previous versions with them
 	if ver.IsSameOrAfter(version.From(7, 7, 0)) {
 		NodeTypesTransformLabelName.Set(nodeRoles.HasTransformRole(), labels)
+		NodeTypesRemoteClusterClientLabelName.Set(nodeRoles.HasRemoteClusterClientRole(), labels)
+	}
+	// voting_only master eligible nodes were added only in 7.3.0 so we don't want label prior versions with it
```
Comment (sebgl) [inline pkg/controller/elasticsearch/label/label.go:128]: `we don't want to`?

File: pkg/controller/elasticsearch/settings/canonical_config.go [hunk 9]
```diff
@@ -10,6 +10,18 @@ import (
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common/version"
 )
 
+// replaceNilNodeRoles works around a rendering issue with 0-sized string arrays/slices
+// that have gone through ucfg and have been reduced to nil.
+// Elasticsearch's new node.roles syntax relies on empty string arrays in YAML to express
+// a coordinating only node.
```
Comment (sebgl) [inline pkg/controller/elasticsearch/settings/canonical_config.go:16]: let's link to https://github.com/elastic/cloud-on-k8s/issues/3718 here maybe?

**Review by sebgl — APPROVED** [review]
LGTM

File: test/e2e/es/topology_test.go [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+	"github.com/elastic/cloud-on-k8s/test/e2e/test/elasticsearch"
+)
+
+//TestCoordinatorNodes tests a cluster with coordinator nodes.
+func TestCoordinatorNodes(t *testing.T) {
+	b := elasticsearch.NewBuilder("test-es-coord").
+		WithESMasterDataNodes(1, elasticsearch.DefaultResources).
+		WithESCoordinatorNodes(1, elasticsearch.DefaultResources)
+
+	test.Sequence(nil, test.EmptySteps, b).RunSequential(t)
```
Comment (pebrc) [inline test/e2e/es/topology_test.go:20]: I forgot to `git add` this one

**Review by david-kow — APPROVED** [review]

**Review by sebgl — APPROVED** [review]
LGTM

**Comment by pebrc** [comment]
Jenkins test this please
